apps/omg_child_chain/lib/omg_child_chain/ethereum_event_aggregator.ex

```diff
@@ -0,0 +1,269 @@
+
+  # I've  not quite convinced this is benefical for us in any way.
+  # ETH listeners move at their own speed... But so is Event Fetcher.
+  # Each ETH listner request is providing a range of eth blocks - and since
+  # we gather events in bulk, they'll be here for other ETH listeners as well...
+  #   def handle_info({:internal_event_bus, :ethereum_new_height, new_height_blknum}, state) do
+  #     from_block = new_height_blknum
+  #     to_block = new_height_blknum
+  #     true = delete_old_logs(new_height_blknum, state)
+  #     true = retrieve_and_store_logs(from_block, to_block, state)
+  #     {:noreply, state}
+  #   end
+
+  def handle_call({:in_flight_exit_piggybacked, from_block, to_block}, _, state) do
+    names = [:in_flight_exit_output_piggybacked, :in_flight_exit_input_piggybacked]
+
+    handout_logs =
+      Enum.reduce(names, [], fn name, acc ->
+        signature =
+          state.events
+          |> Enum.find(fn event -> Keyword.fetch!(event, :name) == name end)
+          |> Keyword.fetch!(:signature)
+
+        logs = handout_log(signature, from_block, to_block, state)
+        logs ++ acc
+      end)
+
+    {:reply, {:ok, handout_logs}, state, {:continue, from_block}}
+  end
+
+  def handle_call({name, from_block, to_block}, _, state) do
+    signature =
+      state.events
+      |> Enum.find(fn event -> Keyword.fetch!(event, :name) == name end)
+      |> Keyword.fetch!(:signature)
+
+    logs = handout_log(signature, from_block, to_block, state)
+    {:reply, {:ok, logs}, state, {:continue, from_block}}
+  end
+
+  def handle_continue(new_height_blknum, state) do
+    _ = delete_old_logs(new_height_blknum, state)
+    {:noreply, state}
+  end
+
+  defp retrieve_and_store_logs(from_block, to_block, state) do
+    from_block
+    |> get_logs(to_block, state)
+    |> enrich_logs_with_call_data(state)
+    |> store_logs(from_block, to_block, state)
+  end
+
+  defp get_logs(from_height, to_height, state) do
+    {:ok, logs} = state.rpc.get_ethereum_events(from_height, to_height, state.event_signatures, state.contracts)
+    Enum.map(logs, &Abi.decode_log(&1))
+  end
+
+  # we get the logs from RPC and we cross check with the event definition if we need to enrich them
+  defp enrich_logs_with_call_data(decoded_logs, state) do
+    events = state.events
+    rpc = state.rpc
+
+    Enum.map(decoded_logs, fn decoded_log ->
+      decoded_log_signature = decoded_log.event_signature
+
+      event = Enum.find(events, fn event -> Keyword.fetch!(event, :signature) == decoded_log_signature end)
+
+      case Keyword.fetch!(event, :enrich) do
+        true ->
+          {:ok, enriched_data} = rpc.get_call_data(decoded_log.root_chain_txhash)
+
+          enriched_data_decoded = enriched_data |> from_hex |> Abi.decode_function()
+          Map.put(decoded_log, :call_data, enriched_data_decoded)
+
+        _ ->
+          decoded_log
+      end
+    end)
+  end
+
+  defp store_logs(decoded_logs, from_block, to_block, state) do
```

Comment:
A completely legitimate comment. There's a balance between keeping things inside a module or extracting specific implementations out. Immediately when you extract things out, you're dealing with hoping between modules when you're reading code. That's always a struggle, but you benefit greatly because you can individually test said functions.

When I decided not to do it I asked myself .... what would I gain from extracting it out? I don't think it anything meaningful! Because I've paramatrized the ETS table and the module doing RPC calls, I can assert these underlying private functions behave exactly as they should. Lemme know if this argument persuades you!

Reply:
Since you've thought this through properly, I'm feeling persuaded. I actually missed the parametized `rpc` part which is great. I'd probably still prefer the logic in its own module, but it's just a personal preference (this module is not big or complicated enough for me to push harder on splitting it). And as you said below, we still need to figure out how to split things to reuse them in ch ch + watcher, we can worry about it at that time. 

👍